### PR TITLE
Add batch splitting mode to Gandalf (splits=2)

### PR DESCRIPTION
**Pull request by mduecker**
## Summary

- Adds `max_concurrency` config field (`int | None`, default `None`, `>= 1`) to control parallel judge sessions
- Renames `mode="sequential"` to `mode="individual"` (old value accepted with deprecation warning)
- Removes `--mode` and `--splits` CLI args — grading config is TOML-only
- Each concurrent chunk runs as a separate judge subprocess via `ThreadPoolExecutor`, results merged in original rubric order with LLM usage summed
- If any split fails (crash, timeout, or unhandled exception), all criteria are marked `met=None` → existing hard-fail path writes `info.json` but **not** `reward.json`, preventing silent partial scores

Closes [HPT-111](https://linear.app/joinhandshake/issue/HPT-111/add-batch-splitting-mode-to-gandalf-splits2)

## Design: two orthogonal axes

The config cleanly separates two independent concerns:

| Axis | Config field | What it controls |
|---|---|---|
| **Granularity** | `mode` | How many criteria per judge session: `"individual"` (one each) or `"batch"` (all at once) |
| **Parallelism** | `max_concurrency` | How many judge sessions run simultaneously: `None` (serial) or `N` (parallel) |

These combine as follows:

| | `max_concurrency=None` | `max_concurrency=N` |
|---|---|---|
| **individual** | One criterion per session, serial (previous "sequential" behavior) | N individual sessions running at once |
| **batch** | All criteria in one session (previous "batch" behavior) | Criteria split into N positional chunks, each a separate batch session |

## Usage

```toml
mode = "batch"
max_concurrency = 2
```

Default is `max_concurrency = None` — zero behavior change for existing configs. `mode = "sequential"` is still accepted with a deprecation warning.

## Error handling

Critical safety invariant validated by `test_no_partial_scores_when_split_fails`:
- When one split fails persistently, `reward.json` is **never** written
- `info.json` contains **all** criteria (both succeeded and failed splits)
- Process exits with code 1
- On unexpected exceptions, remaining futures are cancelled via `executor.shutdown(cancel_futures=True)` and usage is reset to `{}` for consistency with the all-error result set (all-or-nothing semantics)

This prevents the silent partial-score bug from the old `auto_split_rubric.py` ([Slack thread](https://joinhandshake.slack.com/archives/C0A9LSJRZ09/p1774474260805669)).

## Edge cases handled

- **Empty rubric**: early return with `[], {}` — no crash from `range(0, 0, 0)`
- **`max_concurrency` > rubric size**: thread pool workers capped to `min(concurrency, n)`
- **`batch_timeout` with concurrency**: applies per chunk independently, not total across all chunks (documented in config docstring and README)

## Test plan

- [x] 151 tests pass (`pytest tests/`)
- [x] Config: `max_concurrency` defaults to None, rejects 0 and negative values
- [x] Config: `mode="sequential"` migrates to `"individual"` with deprecation warning
- [x] Splitting: even splits, uneven splits, splits exceeding rubric size, empty rubric
- [x] Parallel execution: unique trace files per split, correct timeout per chunk
- [x] Error scenarios: one split crash, timeout, all splits fail, unhandled exception, fewer verdicts
- [x] No partial scores regression test (reward.json absent when split fails)
- [x] Dispatch boundary: `max_concurrency=None` routes to `_run_batch` (not `_run_batch_concurrent`)
- [x] Retry logic works on merged results from splits
- [x] Individual mode with concurrency support
- [x] Exception handler returns consistent empty usage (not partial)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/gandalf_grader/__main__.py [hunk 1]
```diff
@@ -474,6 +478,125 @@ def _run_batch(
...
+    back in original rubric order.
+    """
+    n = len(rubric)
+    chunk_size = math.ceil(n / config.batch_splits)
+    chunks: list[list[tuple[int, RubricItem]]] = []
+    for start in range(0, n, chunk_size):
+        chunks.append([(i, rubric[i]) for i in range(start, min(start + chunk_size, n))])
+
+    print(
+        f"[batch-splits] Splitting {n} criteria into {len(chunks)} chunks "
```
Comment (ronakchaudhary-hs) [inline src/gandalf_grader/__main__.py]: If `rubric` is empty (`n=0`), then `chunk_size = math.ceil(0 / batch_splits) = 0`, and `range(0, 0, 0)` raises `ValueError: range() arg 3 must not be zero`.

Admittedly an empty rubric is unlikely in practice, but this would crash with an unhelpful traceback. Easy fix:

```python
if n == 0:
    return [], {}
```

File: src/gandalf_grader/__main__.py [hunk 2]
```diff
@@ -474,6 +478,125 @@ def _run_batch(
...
+        for j, (orig_idx, item) in enumerate(chunk):
+            v = verdicts[j] if j < len(verdicts) else {}
+            result = CriteriaResult(
+                criteria=item.criteria,
+                weight=item.weight,
+                met=v.get("met"),
+                reasoning=v.get("reasoning", "No reasoning provided."),
+                evidence=v.get("evidence", []),
+            )
+            indexed_results.append((orig_idx, result))
```
Comment (ronakchaudhary-hs) [inline src/gandalf_grader/__main__.py:585]: **All-or-nothing exception handler discards successful split results.**

When iterating `futures` in submission order, if split 1 raises after split 0 has already succeeded, the `except` block discards split 0's results and marks *all* criteria as `met=None`. Meanwhile `total_usage` retains partial usage from the successful split — so the returned usage is inconsistent with the results (usage says we spent tokens, results say everything failed).

This is the same safety tradeoff we had in `auto_split_rubric.py` so I understand the intent. Two suggestions:

1. Reset `total_usage = {}` in the except block so usage is consistent with results.
2. Add a brief comment explaining the all-or-nothing choice, since a reader might expect per-split error isolation.

File: src/gandalf_grader/__main__.py [hunk 3]
```diff
@@ -670,7 +807,10 @@ def main() -> None:
 
     # 1. Initial evaluation
     if config.mode == "batch":
```
Comment (ronakchaudhary-hs) [inline src/gandalf_grader/__main__.py:821]: **`batch_splits > 1` is silently ignored when `mode="sequential"`.**

If someone sets `batch_splits=2` in their `verifier.toml` but forgets to also set `mode = "batch"`, nothing happens and there's no feedback. Worth logging a warning:

```python
if config.batch_splits > 1 and config.mode != "batch":
    print(f"[warning] batch_splits={config.batch_splits} ignored because mode={config.mode!r}", file=sys.stderr)
```

File: src/gandalf_grader/config.py [hunk 4]
```diff
@@ -47,6 +51,7 @@ class VerifierConfig(BaseModel):
     judge_guidance_path: str | None = None
     batch_timeout: int | None = None
     mode: Literal["sequential", "batch"] = "sequential"
+    batch_splits: int = Field(default=1, ge=1)
```
Comment (ronakchaudhary-hs) [inline src/gandalf_grader/config.py]: **Consider an upper bound or warning for large `batch_splits` values.**

`ge=1` is the only validation. `batch_splits=1000` on a 5-criteria rubric creates 5 threads (capped by `math.ceil`), which is wasteful but not harmful. A `le=` bound or a runtime warning when `batch_splits > len(rubric)` would be defensive.

Minor — not blocking.

File: src/gandalf_grader/__main__.py [hunk 5]
```diff
@@ -474,6 +478,125 @@ def _run_batch(
...
+            BatchCriterion(index=orig_idx, criteria=item.criteria)
+            for orig_idx, item in chunk
+        ]
+
+        n_criteria = len(criteria_list)
+        batch_timeout = config.judge_timeout * n_criteria
+        if config.batch_timeout is not None:
+            batch_timeout = min(batch_timeout, config.batch_timeout)
+
+        print(
```
Comment (ronakchaudhary-hs) [inline src/gandalf_grader/__main__.py:550]: `_run_split` calls `print()` inside thread workers. CPython's GIL prevents corruption, but output from parallel splits will interleave unpredictably for `splits > 2`. Fine for now with the default `splits=2`, just noting it.

**Review by ronakchaudhary-hs — COMMENTED** [review]
Nice PR — clean additive feature with solid test coverage (677 lines of tests for 145 lines of impl). A few things to tighten up below, mostly edge cases and a potential crash.

File: src/gandalf_grader/__main__.py (hunk 3 above)
Comment (mduecker) [reply src/gandalf_grader/__main__.py:821]: No longer an issue with the latest change to max_concurrency and modes (individual and batch)

File: src/gandalf_grader/__main__.py (hunk 5 above)
Comment (mduecker) [reply src/gandalf_grader/__main__.py:550]: Skipping this for now, since it's a cosmetic fix and would make the PR more complex right now

**Review by ronakchaudhary-hs — APPROVED** [review]
